distro/otel_test.go

```diff
@@ -16,116 +16,527 @@ package distro_test
+
+	return &tls.Config{
+		Certificates: []tls.Certificate{cert},
+		MinVersion:   tls.VersionTLS13,
+	}
+}
+
+func TestRunJaegerExporterTLS(t *testing.T) {
+	reqCh, hFunc := reqHander()
+	srv := httptest.NewUnstartedServer(hFunc)
+	t.Cleanup(srv.Close)
+
+	srv.TLS = serverTLSConfig(t)
+	srv.StartTLS()
+
+	t.Cleanup(distro.Setenv("OTEL_TRACES_EXPORTER", "jaeger-thrift-splunk"))
+	sdk, err := distro.Run(
+		distro.WithEndpoint(srv.URL),
+		distro.WithTLSConfig(clientTLSConfig(t)),
+	)
+	require.NoError(t, err)
+
+	ctx := context.Background()
+	_, span := otel.Tracer("TestRunJaegerExporterTLS").Start(ctx, spanName)
+	span.End()
+
+	// Flush all spans from BSP.
+	require.NoError(t, sdk.Shutdown(ctx))
+
+	got := <-reqCh
+	assert.Equal(t, "application/x-thrift", got.Header.Get("Content-type"))
+	assert.True(t, got.TLS.HandshakeComplete, "did not perform TLS exchange")
+}
+
+func TestRunJaegerExporterDefault(t *testing.T) {
+	reqCh, hFunc := reqHander()
+	srv := httptest.NewUnstartedServer(hFunc)
+	t.Cleanup(srv.Close)
+
+	// Start server at default address.
+	ln, err := net.Listen("tcp", "127.0.0.1:9080")
+	require.NoError(t, err)
+	srv.Listener = ln
+	srv.Start()
+
+	t.Cleanup(distro.Setenv("OTEL_TRACES_EXPORTER", "jaeger-thrift-splunk"))
+	sdk, err := distro.Run()
+	require.NoError(t, err)
+
+	ctx := context.Background()
+	_, span := otel.Tracer("TestRunJaegerExporterDefault").Start(ctx, spanName)
+	span.End()
+
+	// Flush all spans from BSP.
+	require.NoError(t, sdk.Shutdown(ctx))
+
+	got := <-reqCh
+	assert.Equal(t, "application/x-thrift", got.Header.Get("Content-type"))
+}
+
+type exportRequest struct {
+	Header metadata.MD
+	Spans  []*tpb.Span
+}
+
+type collectorTraceServiceServer struct {
+	ctpb.UnimplementedTraceServiceServer
+
+	requests chan exportRequest
+}
+
+var _ ctpb.TraceServiceServer = (*collectorTraceServiceServer)(nil)
+
+func (ctss *collectorTraceServiceServer) Export(ctx context.Context, exp *ctpb.ExportTraceServiceRequest) (*ctpb.ExportTraceServiceResponse, error) {
+	rs := exp.ResourceSpans[0]
+	ils := rs.GetInstrumentationLibrarySpans()[0]
+	headers, _ := metadata.FromIncomingContext(ctx)
+
+	ctss.requests <- exportRequest{Header: headers, Spans: ils.GetSpans()}
+
+	return &ctpb.ExportTraceServiceResponse{}, nil
+}
+
+type collector struct {
+	endpoint     string
+	srv          *grpc.Server
+	traceService *collectorTraceServiceServer
+}
+
+func newCollector(t *testing.T) *collector {
+	coll, err := newCollectorAt("localhost:0")
+	require.NoError(t, err)
+	t.Cleanup(coll.srv.GracefulStop)
+	return coll
+}
+
+func newCollectorAt(address string) (*collector, error) {
+	ln, err := net.Listen("tcp", address)
+	if err != nil {
+		return nil, err
+	}
+
+	coll := &collector{
+		endpoint: ln.Addr().String(),
+		traceService: &collectorTraceServiceServer{
+			requests: make(chan exportRequest, 1),
+		},
+	}
+
+	srv := grpc.NewServer()
+	ctpb.RegisterTraceServiceServer(srv, coll.traceService)
+	go func() { _ = srv.Serve(ln) }()
```

Comment:
you can consider logging the error returned from https://pkg.go.dev/google.golang.org/grpc#Server.Serve

I think it helped me once in the past but I do not remember exactly what was the case...